store-gateway: clean up chunks fetching, deprecate bucketed chunks pool

dimitarvdimitrov opened the pull request:
#### What this PR does

I started this as an internal refactor to make use of the existing `offsetTrackingReader` that #4926 introduced.

But it turned out that I can slightly refactor the chunks refetching logic and get rid of a lot of code (~800 LOC). I also covered it with some tests because it was untested until now. If it's easier to review, I can split this into multiple PRs.

This allowed to remove the method for fetching a single chunk range and all the related code blocks
* `pool.BucketedBytes`
* `storegateway.byteRange`
* `bucketBlock.readChunkRange`

Because of this the following config options are no longer used and can be removed
* `-blocks-storage.bucket-store.max-chunk-pool-bytes`
* `-blocks-storage.bucket-store.chunk-pool-min-bucket-size-bytes`
* `-blocks-storage.bucket-store.chunk-pool-max-bucket-size-bytes`

I marked them as deprecated but they are also unused. I think this falls out of line with our deprecation policy, but presently the options were only used when refetching chunks, which is a very rare scenario and weren't being respected in the regular case.


related to #3939 

#### Checklist

- [x] Tests updated
- [x] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

dimitarvdimitrov commented on pkg/storegateway/bucket_chunk_reader.go:
```diff
@@ -65,7 +67,7 @@ func (r *bucketChunkReader) addLoad(id chunks.ChunkRef, seriesEntry, chunkEntry
 	r.toLoad[seq] = append(r.toLoad[seq], loadIdx{
 		offset:      off,
-		length:      length,
+		length:      util_math.Max(varint.MaxLen32, length), // If the length is 0, we need to at least fetch the length of the chunk.
```
After adding tests, I discovered that the bucketChunkReader cannot handle 0 estimations even though 0 is a valid value according to this godoc

https://github.com/grafana/mimir/blob/14080d085561056f7ef8aa70a5641cb403491c3b/pkg/storegateway/series_refs.go#L184

This wasn't causing bugs because `length` never would have been 0 as it is always set to a non-zero value here
https://github.com/grafana/mimir/blob/14080d085561056f7ef8aa70a5641cb403491c3b/pkg/storegateway/series_refs.go#L973-L988

pracucci commented on pkg/storegateway/bucket_chunk_reader.go:
```diff
@@ -126,88 +131,66 @@ func (r *bucketChunkReader) loadChunks(ctx context.Context, res []seriesChunks,
 
-		r.block.chunkPool.Put(nb)
+func (r *bucketChunkReader) fetchChunkRemainder(ctx context.Context, seq int, offset, length int64, dest []byte, localStats *queryStats) error {
+	refetchReader, err := r.block.chunkRangeReader(ctx, seq, offset, length)
```
Here we expect that the `dest` length is equal to the input `length`. If it's not, we'll have a silent bug because `io.ReadFull()` ignores `length` and just fill `dest`. For this reason, I would add a check at the beginning of the function to ensure consistency between `len(dest)` and `length`.

pracucci commented on pkg/storegateway/bucket_chunk_reader.go:
```diff
@@ -126,88 +131,66 @@ func (r *bucketChunkReader) loadChunks(ctx context.Context, res []seriesChunks,
-		if len(*nb) != chunkLen {
-			return errors.Errorf("preloaded chunk too small, expecting %d", chunkLen)
+		// We ignore the crc32 after the chunk data.
+		chunkEncDataLen := int(chunkDataLen) + chunks.ChunkEncodingSize
```
[nit] Really a nit, but I would sort them by the same order of their fields in the serialised data.

```suggestion
		chunkEncDataLen := chunks.ChunkEncodingSize + int(chunkDataLen)
```

pracucci commented on pkg/storegateway/bucket_chunk_reader.go:
```diff
@@ -126,88 +131,66 @@ func (r *bucketChunkReader) loadChunks(ctx context.Context, res []seriesChunks,
 		}
 		localStats.chunksTouched++
-		localStats.chunksTouchedSizeSum += chunkLen + crc32.Size
+		localStats.chunksTouchedSizeSum += chunkEncDataLen
```
[nit] Previously we were also counting the varbit length and `crc32.Size`.

pracucci replied:
I vaguely remember something: was the CRC added by you because you found out that for some special chunks the 4 bytes of the CRC was a significant % of the whole chunk data? 🤔 

If we don't count the varbit and CRC length, could this be misleading when computing the "overfetching"?

dimitarvdimitrov replied:
yeah that makes sense. The crc is also accounted for here https://github.com/grafana/mimir/blob/14080d085561056f7ef8aa70a5641cb403491c3b/pkg/storegateway/series_chunks.go#L653

so it makes sense to also count it in `loadChunks`

pracucci replied:
Shouldn't we also add `varint.UvarintSize(uint64(dataLen))` for the same reason?

pracucci reviewed (approved):
Great job, LGTM! The `bucketChunkReader.loadChunks()` is easier to follow now 👍  I just left few minor comments.

> I marked them as deprecated but they are also unused. I think this falls out of line with our deprecation policy, but presently the options were only used when refetching chunks, which is a very rare scenario and weren't being respected in the regular case.

Not a problem to me.

dimitarvdimitrov commented:
thanks for the review @pracucci. I rebased after merging #4995 and addressed most of your comments. The one left is for whether or not we include the crc32 in the `touched` metrics.

pracucci commented on docs/sources/mimir/configure/about-versioning.md:
```diff
@@ -137,7 +137,11 @@ The following features are currently deprecated and will be **removed in Mimir 2
+  - `-blocks-storage.bucket-store.chunk-pool-max-bucket-size-bytes`
+  - `-blocks-storage.bucket-store.max-chunk-pool-bytes`
 - Ingester
   - `-blocks-storage.tsdb.max-tsdb-opening-concurrency-on-startup`
```
This will be removed in 2.10 because announced in 2.8. You should just move the store-gateway config options under 2.11 removals.

dimitarvdimitrov commented:
with this I will also remove the flag for chunks pool from the mimir jsonnet. I pushed this commit to do it https://github.com/grafana/mimir/pull/4996/commits/ebe56c5c00b73cce04ffd330153456c4f10ece00

dimitarvdimitrov commented:
there is a strange 1k loc diff in the helm manifests. I will take a look later

dimitarvdimitrov commented:
fixed. It was a matter of removing a local file and regenerating helm tests

pracucci commented:
Still LGTM. I just left a final comment: https://github.com/grafana/mimir/pull/4996#discussion_r1195395763
